src/coreclr/src/vm/gcenv.os.cpp

```diff
@@ -957,6 +953,54 @@ uint64_t GCToOSInterface::GetPhysicalMemoryLimit(bool* is_restricted)
     return memStatus.ullTotalPhys;
 }
 
+// Set the total physical memory that this process is allowed to use.
+// Remarks:
+//  A process can use a GC config (GCTotalPhysicalMemory) to specify the amount of physicla memory
+//  it's allowed to use. And if no hardlimit (specified with the GCHeapHardLimit config) is specified,
+//  this is treated as the restricted environment so the hardlimit will be calcuated accordingly, ie,
+//  max (75% of the total physical memory, 20mb).
+bool GCToOSInterface::SetRestrictedPhysicalMemoryLimit(uint64_t totalPhysicalMemory)
+{
+    LIMITED_METHOD_CONTRACT;
+    HINSTANCE hinstKernel32 = 0;
+
+#ifndef TARGET_UNIX
+    MEMORYSTATUSEX ms;
+    GetProcessMemoryLoad(&ms);
+    uint64_t total_virtual = ms.ullTotalVirtual;
+
+    if (total_virtual < totalPhysicalMemory)
+    {
+        g_UseRestrictedVirtualMemory = true;
+        totalPhysicalMemory = total_virtual;
+    }
+
+    if (!g_UseRestrictedVirtualMemory)
+    {
+        hinstKernel32 = WszLoadLibrary(L"kernel32.dll");
+        if (!hinstKernel32)
+            goto exit;
+
+        GCGetProcessMemoryInfo = (PGET_PROCESS_MEMORY_INFO)GetProcAddress(hinstKernel32, "K32GetProcessMemoryInfo");
+
+        if (!GCGetProcessMemoryInfo)
+            goto exit;
+    }
+#endif //!TARGET_UNIX
+
+    VolatileStore(&g_RestrictedPhysicalMemoryLimit, (size_t)totalPhysicalMemory);
```

Comment:
This makes the value mutable. It makes the GC/OS contract hard to reason about.

Would it make more sense to change `GCToOSInterface::GetMemoryStatus` to return memory load in absolute bytes? The GC can then re-compute the load against whatever configured limits it is applying.

Reply:
it does make the value mutable but it's only mutable during initialization; and only via this API.

but I can see the merit of what you suggested. I don't have a strong opinion on this since it's fairly subjective. on one hand it's good to make things simple (we already hide things in the OS layer instead of exposing all the details) which makes it easy to reason about on the GC side; on the other hand I can also see the flexibility it provides for the GC side. 

@cshung do you have an opinion?

Reply:
I think a clean design is to make `GCtoOSInterface::Initialize()` take an additional parameter and set the field during initialization. This would do exactly the same as the proposed change does and doesn't have a mutable interface.

Reply:
wouldn't that make `GCtoOSInterface::Initialize()` become a mutable interface exactly like the proposed API? 

honestly it makes no difference to me. both mutate a value on the vm side. instead of adding an API we modify an existing API. but if you actually feel strongly about it, I could go with the latter.

Reply:
Here is my thought:

I think the idea is that we expect `GCtoOSInterface` is an interface for the GC to consume OS capabilities, thus we do not expect the GC to modify its behavior.

Initialization is usually excused for being mutating, as it is expected to gather information (e.g. the function pointers to kernel32.dll functions) before it can function properly. Thus getting the configuration value and restrict the memory usage sounds proper there.

Reply:
I asked @cshung and he doesn't have a strong opinion on this either. @jkotas, I could either go with my current API or if you do prefer Andrew's suggestion as well I can change to that. but I wouldn't want to spend any more time on something like this and I do want to get this in soon since the perf infra work Ivan is doing needs this.

Reply:
Here is what the change would look like without caching the limit in the OS PAL: https://github.com/dotnet/runtime/pull/36152

I have not gone all the way to return absolute bytes from GetMemoryStatus like I have suggested above, but it is going in that direction.